examples/dnn_dcgan_train_ex.cpp

```diff
@@ -0,0 +1,273 @@
+        assign_image_scaled(image, output);
+        images.push_back(std::move(image));
+    }
+    return images;
+}
+
+int main(int argc, char** argv) try
+{
+    // This example is going to run on the MNIST dataset.
+    if (argc != 2)
+    {
+        cout << "This example needs the MNIST dataset to run!" << endl;
+        cout << "You can get MNIST from http://yann.lecun.com/exdb/mnist/" << endl;
+        cout << "Download the 4 files that comprise the dataset, decompress them, and" << endl;
+        cout << "put them in a folder.  Then give that folder as input to this program." << endl;
+        return EXIT_FAILURE;
+    }
+
+    // MNIST is broken into two parts, a training set of 60000 images and a test set of 10000
+    // images.  Each image is labeled so that we know what hand written digit is depicted.
+    // These next statements load the dataset into memory.
+    std::vector<matrix<unsigned char>> training_images;
+    std::vector<unsigned long>         training_labels;
+    std::vector<matrix<unsigned char>> testing_images;
+    std::vector<unsigned long>         testing_labels;
+    load_mnist_dataset(argv[1], training_images, training_labels, testing_images, testing_labels);
+
+    // Fix the random generator seeds for network initialization and noise
+    srand(314159);
+    dlib::rand rnd(std::rand());
+
+    // Instantiate both generator and discriminator
+    generator_type generator;
+    discriminator_type discriminator(
+        leaky_relu_(0.2), leaky_relu_(0.2), leaky_relu_(0.2));
+    // Remove the bias learning from the networks
+    visit_layers(generator, visitor_no_bias());
+    visit_layers(discriminator, visitor_no_bias());
+    // Forward random noise so that we see the tensor size at each layer
+    discriminator(generate_image(generator, make_noise(rnd)));
+    cout << "generator" << endl;
+    cout << generator << endl;
+    cout << "discriminator" << endl;
+    cout << discriminator << endl;
+
+    // The solvers for the generator network.  In this case, we don't need to use a dnn_trainer
+    // for the generator, since we are going to train it manually
+    std::vector<adam> g_solvers(generator.num_computational_layers, adam(0, 0.5, 0.999));
+    auto g_sstack = make_sstack(g_solvers);
+    double learning_rate = 2e-4;
+    // The discriminator trainer
+    dnn_trainer<discriminator_type, adam> d_trainer(discriminator, adam(0, 0.5, 0.999));
+    d_trainer.be_quiet();
+    d_trainer.set_learning_rate(learning_rate);
+    d_trainer.set_learning_rate_shrink_factor(1);
+    cout << d_trainer << endl;
+
+    // Resume training from last sync file
+    if (file_exists("dcgan_sync"))
+    {
+        deserialize("dcgan_sync") >> generator >> d_trainer;
+    }
+
+    const size_t minibatch_size = 64;
+    const std::vector<float> real_labels(minibatch_size, 1.f);
+    const std::vector<float> fake_labels(minibatch_size, -1.f);
+    dlib::image_window win;
+    size_t iteration = 0;
+    while (iteration < 50'000)
```

Comment:
Using ' in integer literals requires C++14, but dlib only requires C++11.  This didn't compile for me when I tried it, I'm not sure why travis is not failing on this, it should be enforcing C++11 :(

Anyway, I want to switch to C++14, but maybe next year.  In my experience it takes 6-7 years before a large enough majority of dlib users upgrade their compilers to allow dlib itself to begin using new compiler features.

Reply:
Ok, it would be really cool to have some "new" C++ features in dlib. However, if I remember it correctly, you added C++11 as a requirement when you added the deep learning API, which was in 2016 (5 years after C++11). Now it's already 6 years after C++14. But I think, at that point it was more of a necessity, and now it would just be for the fancy stuff :)